tests/async_tests/test_connection_pool.py

```diff
@@ -0,0 +1,179 @@
+from typing import AsyncIterator, Tuple
+
+import pytest
+
+import httpcore
+from httpcore._async.base import ConnectionState
+from httpcore._types import URL, Headers
+
+
+class MockConnection(object):
+    def __init__(self, http_version):
+        self.origin = (b"http", b"example.org", 80)
+        self.state = ConnectionState.PENDING
+        self.is_http11 = http_version == "HTTP/1.1"
+        self.is_http2 = http_version == "HTTP/2"
+        self.stream_count = 0
+
+    async def arequest(
+        self,
+        method: bytes,
+        url: URL,
+        headers: Headers = None,
+        stream: httpcore.AsyncByteStream = None,
+        ext: dict = None,
+    ) -> Tuple[int, Headers, httpcore.AsyncByteStream, dict]:
+        self.state = ConnectionState.ACTIVE
+        self.stream_count += 1
+
+        async def on_close():
+            self.stream_count -= 1
+            if self.stream_count == 0:
+                self.state = ConnectionState.IDLE
+
+        async def aiterator() -> AsyncIterator[bytes]:
+            yield b""
+
+        stream = httpcore.AsyncIteratorByteStream(
+            aiterator=aiterator(), aclose_func=on_close
+        )
+
+        return 200, [], stream, {}
+
+    async def aclose(self):
+        pass
+
+    def info(self) -> str:
+        return str(self.state)
+
+    def mark_as_ready(self) -> None:
+        self.state = ConnectionState.READY
+
+    def is_connection_dropped(self) -> bool:
+        return False
+
+
+class ConnectionPool(httpcore.AsyncConnectionPool):
+    def __init__(self, http_version: str):
+        super().__init__()
+        self.http_version = http_version
+        assert http_version in ("HTTP/1.1", "HTTP/2")
+
+    def _create_connection(self, **kwargs):
+        return MockConnection(self.http_version)
+
+
+async def read_body(stream: httpcore.AsyncByteStream) -> bytes:
+    try:
+        body = []
+        async for chunk in stream:
+            body.append(chunk)
+        return b"".join(body)
+    finally:
+        await stream.aclose()
+
+
+@pytest.mark.trio
+async def test_sequential_requests_h11() -> None:
+    async with ConnectionPool(http_version="HTTP/1.1") as http:
```

Comment:
We could parametrize `http_version` since both tests `test_sequential_requests_h11` and `test_sequential_requests_h2` are identical otherwise.

Reply:
Data driven testing :heart:

Reply:
👍